### DISCUSSION TITLE
MobX doesn't triggers re-render after update in fetch's then callback

### DISCUSSION THREAD

**Discussion by hitriyvalenok** [mobxjs/mobx#3091] [Q&A]
Hello there! I am trying to update state from callback in _then_ method. State is updating correctly, but it doesn't re-renders in mobx-react starting from 7 version.

I made an [reproduction example](https://codesandbox.io/s/mobx-doesnt-trigger-re-render-after-update-in-fetchs-then-callback-93rqi). After script execution you should be "authorized", but it's doesn't happen. But if you downgrade mobx-react version to 6.3.1, all will be ok.

Please, say me, how can I manage this? Thank you much!

**Comment by BATCOH** [comment]
Since mobx v6 you should call makeObservable/makeAutoObservable, even with decorators.

[Fixed example](https://codesandbox.io/s/mobx-doesnt-trigger-re-render-after-update-in-fetchs-then-callback-forked-juqwy?file=/src/index.tsx)
[Docs](https://mobx.js.org/enabling-decorators.html)

Also, your [previous question here](https://github.com/mobxjs/mobx/discussions/2938) is about the same type of problem :) Just noting, no judgment.

**Reply by hitriyvalenok** [reply]
> Also, your [previous question here](https://github.com/mobxjs/mobx/discussions/2938) is about the same type of problem :) Just noting, no judgment.

Yeah, I remembered too :)

But unfortunately this isn't my case. I just forgot to add this in the example, but my real code hasn't this problem, but still doesn't works. I am working on preact, not react, maybe there is a problem? My project not tiny and other parts relating to mobx and react works great. A [similar topic](https://stackoverflow.com/questions/51455986/mobx-dont-update-react-dom-in-fetch-promise-callback).

Now I am trying to reproduce my trouble with preact.

**Comment by hitriyvalenok** [comment]
This is simplified source. I would be grateful if you check this. Maybe there is a simple type error which sort of I like to make sometimes :D

**App.js:**
```javascript
import { createElement, Component, Fragment } from 'preact';
import PropTypes from "prop-types";
import { action, makeObservable, observable, runInAction } from "mobx";
import { observer } from "mobx-react";
import API from "./API";

@observer
export default class RushFiles extends Component
{
	@observable isUserAuthorized = false

	/**
	 * @type {API}
	 */
	api

	constructor( props )
	{
		super( props )
		makeObservable( this )

		this.isAuthorized = this.isAuthorized.bind( this )

		this.api = new API()

		setInterval(() => {
			console.log( this.isUserAuthorized ) // will be true
		}, 500)
		this.isAuthorized()
	}

	@action
	isAuthorized()
	{
		this.api.isAuthorized( action( ( username ) => {
			this.isUserAuthorized = true
		} ) )
	}

	render()
	{
		return (
			<Fragment>
				{this.isUserAuthorized ? (
					<div>authorized</div>
				) : (
					<div>NOT authorized</div>
				)}
			</Fragment>
		)
	}
}

RushFiles.propTypes = {
	onSubmit: PropTypes.func.isRequired
}
```

**API.js:**
```javascript
import { API_URL } from "../../constants";

export default class API
{
	isAuthorized( onSuccess )
	{
		this.#request( 'is_authorized', {}, ( response ) => {
			onSuccess()
		} )
	}

	/**
	 * @param {string} target
	 * @param {object} fields
	 * @param {function( object )} onSuccess
	 */
	#request( target, fields, onSuccess )
	{
		fetch( API_URL + '/services/' + target,
			{
				method: 'post',
				headers: {
					'Accept': 'application/json',
					'Content-Type': 'application/json'
				},
				body: JSON.stringify( fields )
			} )
			.then( response => {
				if( response.status === 200 ) {
					return response.json()
				}
				else if( response.status === 401 ) {
					throw new APIUnauthorizedError()
				}
				else {
					throw new Error( response.statusText )
				}
			} )
			.then( json => {
				onSuccess( json ) // callback call
			} )
	}
}

export class APIUnauthorizedError extends Error
{
	constructor( message )
	{
		super( message )
		this.name = "APIUnauthorizedError"
	}
}
```

**Reply by BATCOH** [reply]
I haven't checked, but it looks like here:

```javascript
this.api.isAuthorized( action( ( username ) => {
```
you created a new (inner) action. But it's just an annotation and it shouldn't properly work without makeObservable (yep) call after.
You can define separated method with an action annotation, like `setAuthorized` in first example. and use it in the callback.
```javascript
@action
setAuthorized(bool) {
  this.isAuthorized = bool;
}
...
this.api.isAuthorized(( username ) => {
  this.setAuthorized(true);
});
```

Or you can use `runInAction` instead:

```javascript
this.api.isAuthorized((username) => {
    runInAction(() => {
        this.isUserAuthorized = true
    });
});
```

**Comment by hitriyvalenok** [comment]
There is something with calling onSuccess in _then_. If I just to call _onSuccess()_ in _isAuthorized_, there is no troubles! I see only variant in using async/await to avoid any callbacks and just return the result synchronously, but this is factually a crutch

**Comment by hitriyvalenok** [comment]
I've tried this too :(

**Comment by hitriyvalenok** [comment]
Tried this all again with no result :( State changes, but re-render doesn't triggered
